crates/core/tedge_api/src/address.rs

```diff
@@ -0,0 +1,349 @@
+use std::{marker::PhantomData, time::Duration};
+
+use tokio::sync::mpsc::error::{SendTimeoutError, TrySendError};
+
+use crate::plugin::Message;
+
+#[doc(hidden)]
+pub type AnySendBox = Box<dyn std::any::Any + Send>;
+
+#[doc(hidden)]
+#[derive(Debug)]
+pub struct InternalMessage {
+    pub(crate) data: AnySendBox,
+    pub(crate) reply_sender: tokio::sync::oneshot::Sender<AnySendBox>,
+}
+
+/// THIS IS NOT PART OF THE PUBLIC API, AND MAY CHANGE AT ANY TIME
+#[doc(hidden)]
+pub type MessageSender = tokio::sync::mpsc::Sender<InternalMessage>;
+
+/// THIS IS NOT PART OF THE PUBLIC API, AND MAY CHANGE AT ANY TIME
+#[doc(hidden)]
+pub type MessageReceiver = tokio::sync::mpsc::Receiver<InternalMessage>;
+
+/// An address of a plugin that can receive messages a certain type of messages
+///
+/// An instance of this type represents an address that can be used to send messages of a
+/// well-defined type to a specific plugin.
+/// The `Address` instance can be used to send messages of several types, but each type has to be
+/// in `RB: ReceiverBundle`.
+pub struct Address<RB: ReceiverBundle> {
+    _pd: PhantomData<fn(RB)>,
+    sender: MessageSender,
+}
+
+impl<RB: ReceiverBundle> std::fmt::Debug for Address<RB> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct(&format!("Address<{}>", std::any::type_name::<RB>()))
+            .finish_non_exhaustive()
+    }
+}
+
+impl<RB: ReceiverBundle> Clone for Address<RB> {
+    fn clone(&self) -> Self {
+        Self {
+            _pd: PhantomData,
+            sender: self.sender.clone(),
+        }
+    }
+}
+
+impl<RB: ReceiverBundle> Address<RB> {
+    /// THIS IS NOT PART OF THE PUBLIC API, AND MAY CHANGE AT ANY TIME
+    #[doc(hidden)]
+    pub fn new(sender: MessageSender) -> Self {
+        Self {
+            _pd: PhantomData,
+            sender,
+        }
+    }
+
+    /// Send a message `M` to the address represented by the instance of this struct and wait for
+    /// them to accept it
+    ///
+    /// This function can be used to send a message of type `M` to the plugin that is addressed by
+    /// the instance of this type.
+    ///
+    /// # Return
+    ///
+    /// The function either returns `Ok(())` if sending the message succeeded,
+    /// or the message in the error variant of the `Result`: `Err(M)`.
+    ///
+    /// The error is returned if the receiving side (the plugin that is addressed) does not receive
+    /// messages anymore.
+    ///
+    /// # Details
+    ///
+    /// This function may block indefinitely if the receiving end does not start correctly. If this
```

Comment:
This comment sounds inaccurate. This function might block if the receiving end has no more capacity and will wait till some room have been made as messages are processed.

Reply:
Indeed, and if the receiving end never does start correctly then the messages are never dequeued and thus this will never have space.